src/lib/navigation/free-navigation-mode.spec.ts

```diff
@@ -2,12 +2,11 @@ import {Component, ViewChild} from '@angular/core';
 import {async, ComponentFixture, fakeAsync, TestBed, tick} from '@angular/core/testing';
 import {ArchwizardModule} from '../archwizard.module';
 import {WizardComponent} from '../components/wizard.component';
-import {FreeNavigationMode} from './free-navigation-mode';
 
 @Component({
   selector: 'aw-test-wizard',
   template: `
-    <aw-wizard navigationMode="free">
+    <aw-wizard [awNavigationMode] navigateForward="allow">
```

Comment:
Existing unit tests do not show the difference between `SemiStrictNavigationMode` and `FreeNavigationMode`. After translating both of them to `"navigateForward"="allow", unit tests still pass.

What is the difference of `FreeNavigationMode` from `SemiStrictNavigationMode`? Does angular-archwizard have to provide both?

Reply:
The difference between `SemiStrictNavigationMode` is equal to the `FreeNavigationMode` with the exception of its behavior for entering completion steps:
- `SemiStrictNavigationMode` only allows you to enter a completion step if **all** previous steps have been finished
- `FreeNavigationMode` allows you to enter the completion step any time you want

I can't really say whether both modes are required. I'm using the `StrictNavigationMode` in my projects and implemented `FreeNavigationMode` and `SemiStrictNavigationMode` because of some feedback from the GitHub community.

In hindsight to the `SemiStrictNavigationMode` and `FreeNavigationMode` difference I believe it is a good idea to add the new `NavigationModeDirective` the allowed entrance modes for completion steps.

Reply:
It seems that `FreeNavigationMode` can be emulated with `"navigateBackward"="allow" "navigateForward"="allow"` and making all non-completion steps optional. To me, this approach seems cleaner than adding another configuration flag or directive.

I have added a corresponding demo page: https://github.com/madoar/angular-archwizard-demo/pull/40/commits/44f8ddf60040e632e08ca302c7e1fefe704cb619

Reply:
I like the idea. The only issue I have with it is that it changes the color of the step indicators to light green, because they are set as optional. Ideally the steps should look like normal wizard steps. Can we fix that by applying custom css to the example?

Reply:
Yes, I believe it is possible. But is it needed?

I believe optional steps should look like optional :) Suppose you have two wizards on your web site: the first with the default linear workflow and some optional steps (1 -> 2 -> optional 3 -> completion), the second configured with free navigation mode (optional 1, optional 2, optional 3, completion). If there is different styling for required and optional steps, would you prefer (as a visitor of this web site) steps of the second wizard to be styled as "1", "2" of the first, or as the "optional 3"?

Reply:
In this specific case I think it is a better idea to make the optional step look like a "normal" step. I think so because you are telling the users in the example that they can achieve the same behavior as the `FreeNavigationMode` with the new version, if they use the provided code:

```html
<p style="margin: 1em 0;">
  Angular-archwizard ≤ 4.0.0 supported <code>FreeNavigationMode</code> (<code>"navigationMode"="free"</code>).<br>
  You can achieve the same behavior by allowing navigation in both direction and making all steps optional:
</p>
```

The problem is that this is not really the case, because the old `FreeNavigationMode` supports this behavior with "normal" non-optional steps, which look different to the user compared to optional steps.

Reply:
As an alternative solution I think it is also ok to inform the user inside the example that the visualization differs compared to the old free navigation mode and if he absolutely requires the same look & feel , he can override the css to achieve the same visualization compared to before.

Reply:
This sounds very reasonable. Will do (see https://github.com/madoar/angular-archwizard-demo/pull/31).